src/rust/grapl-metrics/src/setup_metrics.rs

```diff
@@ -0,0 +1,89 @@
+use std::panic::catch_unwind;
+
+use opentelemetry::{
+    global::shutdown_tracer_provider,
+    metrics::{self,},
+    runtime,
+    sdk::{
+        export::metrics::aggregation::cumulative_temporality_selector,
+        metrics::{
+            controllers::BasicController,
+            selectors,
+        },
+    },
+    Context,
+};
+use opentelemetry_otlp::{
+    ExportConfig,
+    WithExportConfig,
+};
+use tokio::task;
+pub struct GraplMetrics {
+    provider: BasicController,
+    cx: Context,
+}
+
+impl Drop for GraplMetrics {
+    fn drop(&mut self) {
+        // invokes shutdown on all otel span processors
+        shutdown_tracer_provider();
+        self.provider
+            .stop(&self.cx)
+            .expect("Expected metrics provider stop");
+    }
+}
+
+fn ensure_multithreaded_tokio_runtime() -> impl runtime::Runtime {
+    // https://github.com/open-telemetry/opentelemetry-rust/blob/482772f2317e242e6b98bfe04ed42e4dac8cf77b/opentelemetry/src/sdk/trace/span_processor.rs#L182
+    // https://github.com/open-telemetry/opentelemetry-rust/issues/786#issuecomment-1107884339
+
+    // block_in_place() is currently the best heuristic for detecting if the
+    // current runtime is multithreaded or not (it panics if single-thread).
+    // tokio::main is multi by default, tokio::test is single by default.
+    // Ultimately, this function is here primarily to help guide developers
+    // who accidentally call `setup_metrics()` from a default `tokio::test`.
+
+    // This will be better if https://github.com/tokio-rs/tokio/issues/5088
+    // ever gets landed.
+    let err_if_singlethread = catch_unwind(|| task::block_in_place(|| {}));
+    match err_if_singlethread {
+        Ok(_) => runtime::Tokio,
+        Err(_) => panic!(
+            r#"
+            `setup_metrics()` called from a single-threaded Tokio runtime;
```

Comment:
🤔 It's a little odd for assumptions about where the function is called from to be embedded in this message.

Maybe it's better to just embed this into `setup_metrics` directly?

Reply:
Yes, I could inline it, but I'd much prefer to not include this kind-of-gross-mess in an otherwise straightforward function about setting up metrics. 

It's not a public function, and its entire purpose is to assert things about the runtime it's currently running in. 

Keep in mind: this function is here as a guard rail so other people don't get stuck down the 4-day "what the heck is going wrong" hole I got down in, so perhaps I'm being a bit too precious about this